spring-ws-core/src/main/java/org/springframework/ws/client/core/observation/WebServiceObservationInterceptor.java

```diff
@@ -0,0 +1,185 @@
+    private SAXParser saxParser;
+
+    private WebServiceTemplateConvention customConvention;
+
+    public WebServiceObservationInterceptor(ObservationRegistry observationRegistry) {
+        this.observationRegistry = observationRegistry;
+        parserFactory = SAXParserFactory.newNSInstance();
+        try {
+            saxParser = parserFactory.newSAXParser();
+        } catch (ParserConfigurationException e) {
+            throw new RuntimeException(e);
+        } catch (SAXException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+
+    @Override
+    public boolean handleRequest(MessageContext messageContext) throws WebServiceClientException {
+
+        TransportContext transportContext = TransportContextHolder.getTransportContext();
+        HeadersAwareSenderWebServiceConnection connection =
+                (HeadersAwareSenderWebServiceConnection) transportContext.getConnection();
+
+
+        Observation observation = WebServiceTemplateObservationDocumentation.WEB_SERVICE_TEMPLATE.start(
+                customConvention,
+                DEFAULT_CONVENTION,
+                () -> new WebServiceTemplateObservationContext(connection),
+                observationRegistry);
+
+        messageContext.setProperty(OBSERVATION_KEY, observation);
+
+        return true;
+    }
+
+    @Override
+    public void afterCompletion(MessageContext messageContext, Exception ex) throws WebServiceClientException {
+
+        Observation observation = (Observation) messageContext.getProperty(OBSERVATION_KEY);
+        Assert.notNull(observation, "Expected observation in messageContext");
+
+        WebServiceTemplateObservationContext context = (WebServiceTemplateObservationContext) observation.getContext();
+
+        WebServiceMessage request = messageContext.getRequest();
+        WebServiceMessage response = messageContext.getResponse();
+
+        if (request instanceof SoapMessage soapMessage) {
+
+            Source source = soapMessage.getSoapBody().getPayloadSource();
+            QName root = getRootElement(source);
+            if (root != null) {
+                context.setLocalPart(root.getLocalPart());
+                context.setNamespace(root.getNamespaceURI());
+            }
+            if (soapMessage.getSoapAction() != null && !soapMessage.getSoapAction().equals(TransportConstants.EMPTY_SOAP_ACTION)) {
+                context.setSoapAction(soapMessage.getSoapAction());
+            }
+        }
+
+        if (response instanceof FaultAwareWebServiceMessage faultAwareResponse) {
+            if (!faultAwareResponse.hasFault() && ex == null) {
+                context.setOutcome("success");
+            } else {
+                context.setOutcome("fault");
+            }
+        }
+
+        URI uri = getUriFromConnection();
+        if (uri != null) {
+            context.setHost(uri.getHost());
+
+            context.setContextualName("POST " + uri.getPath());
```

Comment:
Could this path be high cardinality? I suppose by convention that isn't usually expected with SOAP endpoints, but if this might contain a path variable, we may want to consider a way to resolve a templated path instead of the raw path.

Reply:
Good point!
As far as my experience goes, I've never seen a WebService published with a variable path but it could certainly be the case.
The quick fix could then be to add it to the HighCardinalityKeys, I'm not sure how to hold of a templated path.